### PR TITLE
Two-level tasks and sampling functions

### PR DESCRIPTION
**Pull request by apoms**
The runtime no longer attempts to automatically elide warmup rows or break tasks into IO items. Each task now explicitly declares the boundaries of "IO items" and how many warmup rows to provide.

Sampling functions allow for a compact description of common patterns (like all, strided, range, etc).

### CODE REVIEW

File: python/scannerpy/sampler.py [hunk 1]
```diff
@@ -17,8 +17,26 @@ def _convert_collection(self, videos):
         else:
             return videos
 
-    def all(self, videos):
-        return self.strided(videos, 1)
+    def all(self, videos, item_size=1000, warmup_size=0):
+        sampler_args = self._db._metadata_types.AllSamplerArgs()
```
Comment (willcrichton) [inline python/scannerpy/sampler.py]: All protobufs should be accessed via `db.protobufs.WhateverArgs` for consistency.

File: python/scannerpy/sampler.py [hunk 2]
```diff
@@ -71,6 +98,16 @@ def strided_range(self, video, start, end, stride):
...
-        sample.rows.extend(
-            range(min(start, num_rows), min(end, num_rows), stride))
+        sample.sampling_function = "StridedRange"
+        sampler_args = self._db._metadata_types.StridedRangeSamplerArgs()
+        sampler_args.stride = stride
+        s = start
+        while s < end:
+            ws = max(0, s - warmup_size * stride)
+            e = min(s + item_size * stride, end)
+            sampler_args.warmup_starts.append(ws)
```
Comment (willcrichton) [inline python/scannerpy/sampler.py]: I'm assuming the runtime will check if the warmup index is the same as the start index and ignore it in that case?

File: python/scannerpy/table.py [hunk 3]
```diff
@@ -37,13 +37,14 @@ def columns(self, index=None):
         return columns[index] if index is not None else columns
 
     def num_rows(self):
-        return self._descriptor.num_rows
+        return self._descriptor.end_rows[-1]
```
Comment (willcrichton) [inline python/scannerpy/table.py]: Why would the number of rows in the table be the same as the last row? If my table had rows [0, 100, 200], then it doesn't have 200 rows.

File: python/scannerpy/table.py [hunk 4]
```diff
@@ -37,13 +37,14 @@ def columns(self, index=None):
...
 
     def rows(self):
         if self._job is None:
             return list(range(self.num_rows()))
         else:
             if len(self._task.samples) == 1:
+                assert(False)
```
Comment (willcrichton) [inline python/scannerpy/table.py]: Why is this `assert(False)`?

File: scanner/engine/master.cpp [hunk 5]
```diff
@@ -215,218 +226,262 @@ class MasterImpl final : public proto::Master::Service {
...
-    if (next_io_item_to_allocate_ < num_io_items_) {
-      io_item->set_item_id(next_io_item_to_allocate_);
-      ++next_io_item_to_allocate_;
-      i32 items_left = num_io_items_ - next_io_item_to_allocate_;
-      if (items_left % 10 == 0) {
-        VLOG(1) << "IO items remaining: " << items_left;
+  grpc::Status NextWork(grpc::ServerContext *context,
+                        const proto::NodeInfo *node_info,
+                        proto::NewWork *new_work) {
+    if (samples_left_ <= 0) {
```
Comment (willcrichton) [inline scanner/engine/master.cpp]: We probably need to wrap this all in a mutex, right?

File: scanner/engine/master.cpp [hunk 6]
```diff
@@ -215,218 +226,262 @@ class MasterImpl final : public proto::Master::Service {
...
-      validate_task_set(meta, job_params->task_set(), job_result);
-      if (!job_result->success()) {
-        // No database changes made at this point, so just return
-        return grpc::Status::OK;
-      }
+    // Write out database metadata so that workers can read it
+    write_job_metadata(storage_, JobMetadata(job_descriptor));
+
+    // Setup initial task sampler
+    task_result_.set_success(true);
```
Comment (willcrichton) [inline scanner/engine/master.cpp]: Why is it necessary to set up a task sampler here? Couldn't you set `samples_left_ = 0` and then the logic in `NewWork` will take care of it?

File: scanner/engine/runtime.cpp [hunk 7]
```diff
@@ -14,76 +14,15 @@
...
-          load_sample->add_rows(sample.rows(s));
-        }
-      }
-      load_work_entries.push_back(load_item);
-
-      allocated_rows += rows_to_allocate;
-    }
-    total_rows += rows_in_task;
-  }
+Result create_new_work(const std::map<std::string, TableMetadata> &table_metas,
```
Comment (willcrichton) [inline scanner/engine/runtime.cpp]: What is this function doing now? Ctrl+F says it's not being used anywhere.

File: python/scannerpy/sampler.py (hunk 1 above)
Comment (apoms) [inline python/scannerpy/sampler.py]: I tried that but it didn't seem to work?

File: python/scannerpy/sampler.py (hunk 2 above)
Comment (apoms) [inline python/scannerpy/sampler.py]: Yup!

File: python/scannerpy/sampler.py (hunk 1 above)
Comment (willcrichton) [inline python/scannerpy/sampler.py]: Oh that's the fault of the `database.py`. I'll fix that

File: python/scannerpy/table.py (hunk 3 above)
Comment (apoms) [inline python/scannerpy/table.py]: These rows are relative to the table itself, not what it samples. So if your table had io items of size 1, end rows would be [1, 2, 3].

File: python/scannerpy/table.py (hunk 4 above)
Comment (apoms) [inline python/scannerpy/table.py]: I wasn't sure what you were trying to support here. You wanted to get the rows from the parent table?

File: scanner/engine/master.cpp (hunk 5 above)
Comment (apoms) [inline scanner/engine/master.cpp]: Good point.

File: scanner/engine/runtime.cpp (hunk 7 above)
Comment (apoms) [inline scanner/engine/runtime.cpp]: I forgot to delete it!

File: scanner/engine/master.cpp (hunk 6 above)
Comment (apoms) [inline scanner/engine/master.cpp]: If I set `next_task_ = 0` I think that would work!

File: scanner/engine/evaluate_worker.cpp [hunk 8]
```diff
@@ -74,42 +74,34 @@ void *pre_evaluate_thread(void *arg) {
   while (true) {
     auto idle_start = now();
     // Wait for next work item to process
-    EvalWorkEntry work_entry;
-    args.input_work.pop(work_entry);
 
+    std::tuple<IOItem, EvalWorkEntry> entry;
```
Comment (apoms) [inline scanner/engine/evaluate_worker.cpp]: Agreed.

File: python/scannerpy/table.py (hunk 4 above)
Comment (willcrichton) [inline python/scannerpy/table.py]: Yeah, seems like it. Looks like it's the else clause here that should be `assert(False)`.

**Pull request merged**